### PR TITLE
[SPARK-40822][SQL] Stable derived column aliases

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to change auto-generation of column aliases (the case when an user doesn't assign any alias explicitly). Before the changes, Spark SQL generates such alias from `Expression` but this PR proposes to take the parse tree (output of lexer), and generate an alias using the term tokens from the tree.

New helper function `ParserUtils.toExprAlias` takes a `ParseTree` from `Antlr4`, and converts it to a `String` using the following simple rule:
- Concatenate all terminal nodes of the lexer tree without any gaps.
- Upper case keywords and numeric literals.

For example, the sequence of tokens "1", "in", "(", "1.0d" ")" is converted to the alias "1IN(1.0D)".

By default, the feature is off, and the SQL config `spark.sql.stableDerivedColumnAlias.enabled` allows to enable it.

Closes #39332 

### Why are the changes needed?
To improve user experience with Spark SQL. It is always best practice to name the result of any expressions in a queries select list,  if one plans to reference them later. This yields the most readable results and stability. However, sometimes queries are generated or we’re just lazy and trust in the auto generated names. The problem is that the auto-generated names are produced by pretty printing the expression tree which is, while “generally” readable, not meant to be stable across long durations of time. For example:
```sql
spark-sql> DESC SELECT substring('hello', 5);
substring(hello, 5, 2147483647)	string
```
the auto-generated column alias `substring(hello, 5, 2147483647)` contains not-obvious elements.

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?
By running new test:
```
$ build/sbt "test:testOnly *.ResolveAliasesSuite"
```

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -490,15 +494,17 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
             case c @ Cast(ne: NamedExpression, _, _, _) => Alias(c, ne.name)()
             case e: ExtractValue =>
               if (extractOnly(e)) {
-                Alias(e, toPrettySQL(e))()
+                Alias(e, getAlias(e, optGenAliasFunc))()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:497]: can we check where we use `metaForAutoGeneratedAlias` and see if we should add it here as well?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -669,17 +669,26 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
...
+    }
+  }
+
   override def visitNamedExpressionSeq(
-      ctx: NamedExpressionSeqContext): Seq[Expression] = {
+      ctx: NamedExpressionSeqContext): Seq[(Expression, Option[Expression => String])] = {
     Option(ctx).toSeq
       .flatMap(_.namedExpression.asScala)
-      .map(typedVisit[Expression])
+      .map(ctx => (typedVisit[Expression](ctx), getAliasFunc(toExprAlias(ctx))))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:684]: it seems we can always create `UnresolvedAlias` here. `UnresolvedAlias` is a noop if its child is already a named expression.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala [hunk 3]
```diff
@@ -88,4 +94,46 @@ class ResolveAliasesSuite extends AnalysisTest {
     checkAliasName(t1.select(DateSub(Literal(Date.valueOf("2021-01-18")), Literal(null))),
       "date_sub(DATE '2021-01-18', NULL)")
   }
+
+  test("SPARK-40822: Stable derived column aliases") {
+    withSQLConf(SQLConf.STABLE_DERIVED_COLUMN_ALIAS_ENABLED.key -> "true") {
+      Seq(
+        // Literals
+        "' 1'" -> "' 1'",
+        """"abc"""" -> """"abc"""",
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:103]: It makes me think if we should always use the standard SQL syntax to generate the alias name, instead of using the original SQL string. The string literal accepts two quote symbols and seems better to normalize it when generating alias names. cc @srielau

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala [hunk 4]
```diff
@@ -88,4 +94,46 @@ class ResolveAliasesSuite extends AnalysisTest {
...
+        "interval 3 year 1 month" -> "interval 3 year 1 month",
+        "x'00'" -> "x'00'",
+        // Preserve case
+        "CAST(1 as tinyint)" -> "CAST(1 as tinyint)",
+        // Brackets
+        "getbit(11L, 2 + 1)" -> "getbit(11L,2+1)",
+        "string(int(shiftleft(int(-1), 31))+1)" -> "string(int(shiftleft(int(-1),31))+1)",
+        "map(1, 'a') [ 5 ]" -> "map(1,'a')[5]",
+        // Preserve type
+        "CAST('123.a' AS long)" -> "CAST('123.a'AS long)",
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:116]: this one is not very ideal

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala [hunk 5]
```diff
@@ -88,4 +94,46 @@ class ResolveAliasesSuite extends AnalysisTest {
...
+        // Spaces
+        "'1' = 1" -> "'1'=1",
+        "upper('a') = upper('A')" -> "upper('a')=upper('A')",
+        "FLOOR(5, 0)" -> "FLOOR(5,0)",
+        "-1" -> "-1",
+        "1 in (1.0)" -> "1 in(1.0)",
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:122]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:684]: Just in case, this function handle not only named expressions but others too. For example:
```sql
select 1;
```
is handled by `visitNamedExpressionSeq()` but not by `visitExpressionSeq()`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala (hunk 4 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:116]: yep, what would you propose? Detect all string literals, and handle them as all chars terms?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala (hunk 5 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:122]: > ditto

This is different case from the previous one. How can I distinguish the two cases?
```floor(``` and ```in(```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala (hunk 3 above)
Comment (srielau) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:103]: You mean convert " to '  for String literals? +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:497]: It is used in 2 places, and both places are covered by `getAlias`. Seems not need to add anything.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:103]: How should I distinguish double quoted string literals from double quoted identifiers on this level?

**Comment by MaxGekk** [comment]
@srielau Could you take a look at the PR one more time, please.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala]: I have added such conversion.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -465,7 +465,20 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
...
+   *          For example: a.b.c => c
+   *     2.2. A CAST, or try_cast -> the argument of the cast.
+   *          For example: CAST(c1 AS INT) => c1
+   *     2.3. A map lookup with a literal -> the map key.
+   *          For example: map[5] => 5
+   *     2.4. Squeezing out unwanted whitespace or comments.
+   *          For example: T.c1 + /* test */ foo( 5 ) => T.c1+foo(5)
+   *     2.5. Normalize SQL string literals when ANSI mode is enabled and the SQL config
+   *          `spark.sql.ansi.doubleQuotedIdentifiers` is set to `true`.
+   *          For example: "abc" => 'abc'
```
Comment (srielau) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:481]: This description seems incomplete. I see examples below with \t, \r \n in strings. We preserve them? I'm doubtful that's useful).
Also you seem to have some rules around whitespace.

**Review by srielau — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 6 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Updated the comment.

**Comment by MaxGekk** [comment]
@cloud-fan @srielau Could you take a look at the PR, please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -465,7 +465,23 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
...
+   *          a letter and a digit.
+   *     2.1. Normalize SQL string literals when ANSI mode is enabled and the SQL config
+   *          `spark.sql.ansi.doubleQuotedIdentifiers` is set to `true`.
+   *          For example: "abc" => 'abc'
+   *     2.2. A multipart identifier (column, field, mapkey, ...) -> the right most identifier.
+   *          For example: a.b.c => c
+   *     2.3. A CAST, or try_cast -> the argument of the cast.
+   *          For example: CAST(c1 AS INT) => c1
+   *     2.4. A map lookup with a literal -> the map key.
+   *          For example: map[5] => 5
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:482]: is this really what we want? what was the previous behavior?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 8]
```diff
@@ -490,15 +510,17 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
...
+                Alias(e, getAlias(e, optGenAliasFunc))()
               } else {
-                Alias(e, toPrettySQL(e))(explicitMetadata = Some(metaForAutoGeneratedAlias))
+                Alias(e, getAlias(e, optGenAliasFunc))(
+                  explicitMetadata = Some(metaForAutoGeneratedAlias))
               }
             case e if optGenAliasFunc.isDefined =>
               Alias(child, optGenAliasFunc.get.apply(e))()
-            case l: Literal => Alias(l, toPrettySQL(l))()
+            case l: Literal => Alias(l, getAlias(l, optGenAliasFunc))()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:520]: is `l.sql` a simpler solution?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 9]
```diff
@@ -490,15 +510,17 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
...
             case e =>
-              Alias(e, toPrettySQL(e))(explicitMetadata = Some(metaForAutoGeneratedAlias))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:501]: we don't need to change this line, because the `case e if optGenAliasFunc.isDefined =>` will be hit first, and `optGenAliasFunc` must be none when we reach here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 10]
```diff
@@ -490,15 +510,17 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
             case c @ Cast(ne: NamedExpression, _, _, _) => Alias(c, ne.name)()
             case e: ExtractValue =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:511]: to match multipart identifier, we need to match a chain of `GetStructField` or `GetArrayStructField`

**Comment by srielau** [comment]
I have lost track of the externally described rules. Do you have an updated summary?
Especially since we recently discussed about what to do with whitespace
Also, do we plan to leave this OFF by default? If we don't make this default it will never get adoption because most folk just won't care.

**Comment by MaxGekk** [comment]
> Do you have an updated summary?

Yes, it is very easy to check - just look at the last commit.

> Especially since we recently discussed about what to do with whitespace

? @srielau Please, look at the PR changes first of all.

> Also, do we plan to leave this OFF by default?

Yes, Serge could you look at the PR's description and changes before asking the questions.

**Comment by srielau** [comment]
> > Do you have an updated summary?
> 
> Yes, it is very easy to check - just look at the last commit.
> 
> > Especially since we recently discussed about what to do with whitespace
> 
> ? @srielau Please, look at the PR changes first of all.
> 
> > Also, do we plan to leave this OFF by default?
> 
> Yes, Serge could you look at the PR's description and changes before asking the questions.
You sum up the root cause for a lot of our QA troubles right there....
Please assure that the description of the PR is accurate and up to date (or have a pointer to a google-doc that does the same).
We cannot reciew code against itself. We have to review it against a common understanding on what it is meant to do.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 9 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:501]: Yep, I over-protected myself to catch all cases of `toPrettySQL()`. Let me revert this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 8 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:520]: The case above should catch the cases of auto-generated stable aliases. This line can be reverted back.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala [hunk 11]
```diff
@@ -259,4 +260,48 @@ object ParserUtils {
       }
     }
   }
+
+  /**
+   * Converts the given parse tree to an alias by
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala:265]: `Normalizes the expression parser tree to a SQL string which will be used to generate expression alias`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala [hunk 12]
```diff
@@ -259,4 +260,48 @@ object ParserUtils {
...
+   * - Adding a gap between two terms when both contains only either letters or digits.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala:266]: let's drop this as it can go very complicated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala [hunk 13]
```diff
@@ -259,4 +260,48 @@ object ParserUtils {
...
+   * - Upper casing keywords and numeric literals.
+   * - Converting double quoted string literals to single quoted.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala:268]: this make the function unstable as it relies on a config, let's drop as well.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala [hunk 14]
```diff
@@ -88,4 +94,46 @@ class ResolveAliasesSuite extends AnalysisTest {
...
+  test("SPARK-40822: Stable derived column aliases") {
+    withSQLConf(SQLConf.STABLE_DERIVED_COLUMN_ALIAS_ENABLED.key -> "true") {
+      Seq(
+        // Literals
+        "' 1'" -> "' 1'",
+        """"abc"""" -> """"abc"""",
+        """'\t\n xyz \t\r'""" -> """'\t\n xyz \t\r'""",
+        "1l" -> "1L", "1S" -> "1S",
+        "date'-0001-1-28'" -> "date'-0001-1-28'",
+        "interval 3 year 1 month" -> "INTERVAL3YEAR1MONTH",
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:107]: @cloud-fan As we discussed offline, I removed gaps between terms.

**Review by srielau — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 15]
```diff
@@ -488,16 +477,13 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
...
-              } else {
-                Alias(e, toPrettySQL(e))(explicitMetadata = Some(metaForAutoGeneratedAlias))
-              }
             case e if optGenAliasFunc.isDefined =>
               Alias(child, optGenAliasFunc.get.apply(e))()
             case l: Literal => Alias(l, toPrettySQL(l))()
             case e =>
+              val metaForAutoGeneratedAlias = new MetadataBuilder()
+                .putString("__autoGeneratedAlias", "true")
+                .build()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:486]: To be safe and not change DataFrame behavior, can we not add this metadata if the expression is extractOnly?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 16]
```diff
@@ -2704,7 +2690,7 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
     }
 
     private def trimAlias(expr: NamedExpression): Expression = expr match {
-      case UnresolvedAlias(child, _) => child
+      case ua: UnresolvedAlias => ua.child
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2693]: unnecessary change

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala [hunk 17]
```diff
@@ -88,4 +94,46 @@ class ResolveAliasesSuite extends AnalysisTest {
...
+
+  test("SPARK-40822: Stable derived column aliases") {
+    withSQLConf(SQLConf.STABLE_DERIVED_COLUMN_ALIAS_ENABLED.key -> "true") {
+      Seq(
+        // Literals
+        "' 1'" -> "' 1'",
+        """"abc"""" -> """"abc"""",
+        """'\t\n xyz \t\r'""" -> """'\t\n xyz \t\r'""",
+        "1l" -> "1L", "1S" -> "1S",
+        "date'-0001-1-28'" -> "date'-0001-1-28'",
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:106]: we should make `date` a keyword. Are we going to do it in a followup?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala [hunk 18]
```diff
@@ -88,4 +94,46 @@ class ResolveAliasesSuite extends AnalysisTest {
...
+        "interval 3 year 1 month" -> "INTERVAL3YEAR1MONTH",
+        "x'00'" -> "x'00'",
+        // Preserve case
+        "CAST(1 as tinyint)" -> "CAST(1AStinyint)",
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:110]: ditto, all types should be keyword.

**Review by cloud-fan — APPROVED** [review]
LGTM except a few minor comments

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala (hunk 17 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:106]: I would like to convert types and their literal constructors in a separate PR if you don't mind.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 15 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:493]: done. Please, double check.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala (hunk 17 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:106]: I opened the JIRA https://issues.apache.org/jira/browse/SPARK-42873

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @srielau and @cloud-fan for review.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala (hunk 18 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveAliasesSuite.scala:110]: This PR https://github.com/apache/spark/pull/40565 does this.
